src/workerd/api/encoding.c++

```diff
@@ -464,25 +465,110 @@ jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
   return js.alloc<TextEncoder>();
 }
 
-namespace {
-TextEncoder::EncodeIntoResult encodeIntoImpl(
-    jsg::Lock& js, jsg::JsString input, jsg::BufferSource& buffer) {
-  auto result = input.writeInto(
-      js, buffer.asArrayPtr().asChars(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
-  return TextEncoder::EncodeIntoResult{
-    .read = static_cast<int>(result.read),
-    .written = static_cast<int>(result.written),
-  };
-}
-}  // namespace
+jsg::JsUint8Array TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
+  jsg::JsString str = input.orDefault(js.str());
+
+  if (str.isOneByte(js)) {
+    auto length = str.length(js);
+    // Fast path for one-byte strings (Latin-1). writeOneByte() copies the raw bytes without
+    // flattening the string, which is more efficient than using ValueView. Note that we
+    // allocate `length * 2` bytes because Latin-1 characters 0x80-0xFF need 2 bytes in UTF-8.
+    auto backing =
+        jsg::BackingStore::alloc<v8::Uint8Array>(js, length, jsg::Lock::AllocOption::UNINITIALIZED);
+    str.writeOneByte(
+        js, backing.asArrayPtr<kj::byte>(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
+    auto backingData = reinterpret_cast<const char*>(backing.asArrayPtr<kj::byte>().begin());
+
+    size_t utf8_length = simdutf::utf8_length_from_latin1(backingData, length);
+
+    if (utf8_length == length) {
+      return jsg::JsUint8Array(backing.createHandle(js).As<v8::Uint8Array>());
+    }
+
+    auto backing2 = jsg::BackingStore::alloc<v8::Uint8Array>(
+        js, utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
+    auto written = simdutf::convert_latin1_to_utf8(
+        backingData, length, reinterpret_cast<char*>(backing2.asArrayPtr<kj::byte>().begin()));
+    KJ_DASSERT(backing2.size() == written);
+    return jsg::JsUint8Array(backing2.createHandle(js).As<v8::Uint8Array>());
```

Comment:
In this path, you're now making three distinct whole copies of the data in memory, which is no better than the flattening case. And if the string is already flattened, you're using way more memory now than what V8 does for the same operation. Given that the biggest issues we've seen with `TextEncoder::encode` relate to memory pressure more than CPU, I don't think this is actually helping us much.